### PR TITLE
Week 20 - Karin Nordkvist & Petra Almgren (Auth 🗝)

### PR DESCRIPTION
**Pull request by karinnordkvist**
**Link to our deployed project:**
* Frontend: https://project-auth-by-karin-and-petra.netlify.app/
* Backend: https://auth-by-karin-petra.herokuapp.com/

**My reflections on how this week's project turned out:** 
The project went well and we worked great together! 😄 

**Things I'd like to have clarified or explained in more detail:**
The subject is really big and it would have been super interesting to get to work with authentication even more. 🌟

### CODE REVIEW

File: backend/server.js [hunk 1]
```diff
@@ -1,29 +1,126 @@
...
+    const accessToken = req.header('Authorization');
+    const user = await User.findOne({ accessToken });
+    if (!user) {
+      throw 'User not found!';
+    }
+    req.user = user;
+    next();
+  } catch (err) {
+    res
+      .status(401)
```
Comment (KarolCislowski) [inline backend/server.js:62]: Good error status 👍🏻

File: frontend/src/components/Content.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      <Button onClick={() => handleClick()}>Reveal my secret message</Button>
+      {secretVisible && <Text>{secretMessage}</Text>}
+      <LogOut />
+    </Wrapper>
+  );
+};
+
+// ----------------------------------------------------------------
+
+const Wrapper = styled.div`
```
Comment (KarolCislowski) [inline frontend/src/components/Content.js:44]: Nice that you used styled-components 💟

File: frontend/src/reducers/users.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+      .then((json) => {
+        if (json.message) {
+          dispatch(users.actions.setError(json.message));
+        } else {
+          dispatch(users.actions.logIn(json));
+          dispatch(users.actions.setError(''));
+        }
+      });
+  };
+};
```
Comment (KarolCislowski) [inline frontend/src/reducers/users.js:64]: Reducer is a great choice to store your user data 🥇 but you need to save some of them in the local storage to stop losing your data after refreshing the page.

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
Great project! Your code is nice and clean! 🥇 

But I need to ask you to fix the logging out on refresh issue 🙏🏻 

Good luck 😸

File: backend/server.js [hunk 4]
```diff
@@ -1,29 +1,126 @@
...
-const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/authAPI"
-mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
-mongoose.Promise = Promise
+const mongoUrl = process.env.MONGO_URL || 'mongodb://localhost/authAPI';
+mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true });
+mongoose.Promise = Promise;
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
+const userSchema = new mongoose.Schema({
+  name: {
```
Comment (emma-tech) [inline backend/server.js:13]: Really nice and well structured schema! :)

File: backend/server.js [hunk 5]
```diff
@@ -1,29 +1,126 @@
...
+  try {
+    const accessToken = req.header('Authorization');
+    const user = await User.findOne({ accessToken });
+    if (!user) {
+      throw 'User not found!';
+    }
+    req.user = user;
+    next();
+  } catch (err) {
+    res
```
Comment (emma-tech) [inline backend/server.js:61]: Nice and well structured code 👍

File: backend/server.js [hunk 6]
```diff
@@ -1,29 +1,126 @@
...
+  } else {
+    res.status(400).json({
+      message: `ERROR: Could not log in. Make sure you've entered the correct user details.`,
+    });
+  }
+});
+
+// Private area - to access this endpoint you'll need to send a request with
+// a header with a value of a valid access token.
+app.get('/private', authenticateUser);
```
Comment (emma-tech) [inline backend/server.js:116]: Really cool that you implemented this private endpoint as well 👍

File: backend/server.js [hunk 7]
```diff
@@ -1,29 +1,126 @@
...
+      message: `ERROR: Could not create user. Make sure you've entered all the fields correctly.`,
+      errors: err,
+    });
+    console.log({ error: err });
+  }
+});
+
+// Log in
+app.post('/sessions', async (req, res) => {
+  const user = await User.findOne({ email: req.body.email });
```
Comment (emma-tech) [inline backend/server.js:104]: Good that you remembered about findOne() here! 🙌

File: backend/server.js [hunk 8]
```diff
@@ -1,29 +1,126 @@
...
+});
+
+// Signup
+app.post('/users', async (req, res) => {
+  try {
+    const { name, email, password } = req.body;
+    const user = new User({ name, email, password });
+    await user.save();
+    res
+      .status(200)
```
Comment (emma-tech) [inline backend/server.js:91]: I was wondering if a 201 status code is more appropriate here, as we are _creating_ and posting a new user? :)

File: backend/server.js [hunk 9]
```diff
@@ -1,29 +1,126 @@
...
+    minlength: [2, 'Oops, the entered name is too short.'],
+  },
+  password: {
+    type: String,
+    required: [true, 'You have to enter a password.'],
+    minlength: [6, 'Oops, the entered password is too short.'],
+  },
+  email: {
+    type: String,
+    required: [true, 'You have to enter an email address.'],
```
Comment (emma-tech) [inline backend/server.js:25]: Good job with validation 👏

**Review by emma-tech — COMMENTED** [review]
Good job Karin and Petra! Really nice and well structured code, and I looove the frontend, so nice and clean styling! 😍

File: frontend/src/components/Main.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+import { useSelector } from 'react-redux';
+
+// Components
+import { SignUp } from './SignUp';
+import { LogIn } from './LogIn';
+import { ErrorMessage } from './ErrorMessage';
+import { Content } from './Content';
+
+// ----------------------------------------------------------------
+
```
Comment (aagneshansson) [inline frontend/src/components/Main.js:12]: Love that you added a specific component for handling whether the user is logged in/not 🙌

**Review by KarolCislowski — APPROVED** [review]
After the discussion with teachers about this assignment, we decided to approve this project without implementing the local storage functionality. 

However for the future, please remember to store your users' authentication data in your browser to make your UX good.

Good luck with your Final Project! 🍀

**Pull request closed without merging**